Process-based steel model (optional alternative implementation)

JakobBD opened the pull request:
## Purpose of this PR

Introduce a process-based implementation for steel production. 
Is only enabled if `cm_subsec_model_steel=processes`. Otherwise, the only difference to `develop` are a few empty sets.
Further implementation details can be found in the in-code documentation and this file: [ImplementationDetails.pdf](https://github.com/remindmodel/remind/files/13979347/ImplementationDetails.pdf)
[Corresponding Remind2 PR](https://github.com/pik-piam/remind2/pull/509)

## Type of change

(Make sure to delete from the Type-of-change list the items not relevant to your PR)

- [x] New feature 
- [x] Fundamental change

## Checklist:

- [x] My code follows the [coding etiquette](https://github.com/remindmodel/remind/blob/develop/main.gms#L80)
- [x] I performed a self-review of my own code
- [x] I explained my changes within the PR, particularly in hard-to-understand areas
- [x] I checked that the [in-code documentation](https://github.com/remindmodel/remind/blob/develop/main.gms#L120) is up-to-date
- [x] I adjusted the reporting in [`remind2`](https://github.com/pik-piam/remind2) where it was needed
- [x] I adjusted `forbiddenColumnNames` in [readCheckScenarioConfig.R](https://github.com/remindmodel/remind/blob/develop/scripts/start/readCheckScenarioConfig.R) in case the PR leads to deprecated switches
- [ ] All automated model tests pass (`FAIL 0` in the output of `make test`) (**note**: did not run make test, but ran all AMT runs, see below)
- [x] The changelog `CHANGELOG.md` [has been updated correctly](https://gitlab.pik-potsdam.de/rse/rsewiki/-/wikis/Standards-for-Writing-a-Changelog)

## Further information (optional):

* Test runs are here: `/p/tmp/jakobdu/remind_pbs/`
* Comparison of results (what changes by this PR?):
  * A comparison of both branches for the ces-basd implementation for all AMT runs can be found here: `/p/tmp/jakobdu/remind_pbs/CompareScenarios_PullRequest/ces__compare_branches/selection`. The neighbouring folder `.../all` has the complete results. It shows that the ces-based model is unaffected by the change. The only considerable differences are in EU21-PkBudget runs, which (also) did not converge in the develop AMT runs. I did not run `make test`, but I did run `Rscript start.R config/scenario_config.csv startgroup=AMT`.
  * A comparison of the CES-based and process-based implementation follows in a reply (see below). Some pdf's are here:  
[FE_mix.pdf](https://github.com/remindmodel/remind/files/13979495/FE_mix.pdf)
[prices.pdf](https://github.com/remindmodel/remind/files/13979496/prices.pdf)
[production.pdf](https://github.com/remindmodel/remind/files/13979497/production.pdf)
[production_by_route.pdf](https://github.com/remindmodel/remind/files/13979498/production_by_route.pdf)
[specific_energy.pdf](https://github.com/remindmodel/remind/files/13979501/specific_energy.pdf)
The full cs2 pdfs are located in `/p/tmp/jakobdu/remind_pbs/CompareScenarios_PullRequest/processes_vs_ces/all`. 

## Note to the reviewers

Since the code changes are substantial, it would be great if I could merge this soon to avoid resolving merge conflicts down the road. 
So for details regarding the process-based implementation, I'd rather open issues and merge those changes within another PR. Any issues in the default (ces-based) configuration should of course be addressed before the merge.

JakobBD commented:
![image](https://github.com/remindmodel/remind/assets/44020564/3361834f-c107-4e63-81b6-606e0b80e555)
![image](https://github.com/remindmodel/remind/assets/44020564/40d76251-d2fc-4007-8df2-8c957017b382)
![image](https://github.com/remindmodel/remind/assets/44020564/5d62f46b-4535-42b5-a368-1613d23633da)
![image](https://github.com/remindmodel/remind/assets/44020564/d51bb610-dfa8-4b6c-95be-469c28ec609b)
![image](https://github.com/remindmodel/remind/assets/44020564/a8a0656c-33c2-4c40-89cd-87605aa23ca1)

JakobBD commented:
FYI @cchrisgong @PhilippVerpoort @mellamoSimon @FalkoUeckerdt

LaviniaBaumstark commented on core/sets.gms:
```diff
@@ -396,16 +405,23 @@ all_enty             "all types of quantities"
-        steel        "Steel"
+        prsteel      "Primary steel"
+        sesteel      "Secondary steel"
+        !!steel        "Steel"
```
why is there still a non active `steel`?  Could it be deleted?

LaviniaBaumstark commented on modules/05_initialCap/on/preloop.gms:
```diff
@@ -19,16 +19,17 @@ s05_inic_switch = 1;
     sum(fe2ppfEn(enty,in),
       pm_cesdata("2005",regi,in,"quantity")
     + pm_cesdata("2005",regi,in,"offset_quantity")
-  ) * s05_inic_switch
```
why is this switch no longer necessary here?

JakobBD replied:
It was on every term separately, I just changed the brackets to be around all the terms and added the switch once for all terms

LaviniaBaumstark commented on modules/05_initialCap/on/preloop.gms:
```diff
@@ -19,16 +19,17 @@ s05_inic_switch = 1;
   + sum(ue2ppfen(enty,in),
       pm_cesdata("2005",regi,in,"quantity")
     + pm_cesdata("2005",regi,in,"offset_quantity")
-    ) * s05_inic_switch
```
same here?

LaviniaBaumstark commented on modules/05_initialCap/on/preloop.gms:
```diff
@@ -42,9 +43,17 @@ q05_eedemini(regi,enty)..
         * pm_shFeCes("2005",regi,enty2,in,teEs2)
         )
       )
-    ) * s05_inic_switch
```
and here?

LaviniaBaumstark commented on modules/05_initialCap/on/preloop.gms:
```diff
@@ -518,6 +537,8 @@ if (cm_startyear gt 2005,
   Execute_Loadpoint 'input_ref' vm_deltaCap.lo = vm_deltaCap.lo;
   Execute_Loadpoint 'input_ref' vm_deltaCap.up = vm_deltaCap.up;
+  !! moved to industry module:
+  !!Execute_Loadpoint 'input_ref' p37_specFeDem = p37_specFeDem;
```
I think, this comment will not be helpful in a few mounths. Delete?

LaviniaBaumstark commented on modules/37_industry/subsectors/datainput.gms:
```diff
@@ -568,4 +550,145 @@ execute_load "input_ref.gdx", vm_demFEsector;
+p37_specFeDemTarget("feels","idr","h2")           = 0.08 / (sm_TWa_2_MWh/sm_giga_2_non);    !! Source: POSTED / Hölling et al 2017, Just before Table 1 (per tHBI)
+
+p37_specFeDemTarget("fegas","idr","ng")           = 2.69 / (sm_TWa_2_MWh/sm_giga_2_non);    !! Source: POSTED / Hölling et al 2017, Page 7 (9.7 GJ) (per tHBI)
+p37_specFeDemTarget("feels","idr","ng")           = 0.08 / (sm_TWa_2_MWh/sm_giga_2_non);    !! Source: POSTED / Hölling et al 2017, Page 7 (9.7 GJ) (per tHBI)
```
this will be a file generated via pre-processign based on POSTED, right?

JakobBD replied:
yes

LaviniaBaumstark commented on modules/37_industry/subsectors/datainput.gms:
```diff
@@ -568,4 +550,145 @@ execute_load "input_ref.gdx", vm_demFEsector;
+  vm_outflowPrc.fx('2005',regi,'bf','standard') = p37_specMatDem("pigiron","bof","unheated") * vm_outflowPrc.l('2005',regi,'bof','unheated');
+  vm_outflowPrc.fx('2005',regi,'eaf','sec') = pm_fedemand('2005',regi,'ue_steel_secondary');
+  vm_outflowPrc.fx('2005',regi,'eaf','pri') = 0.;
+  vm_outflowPrc.fx('2005',regi,'idr','ng') = 0.;
```
sounds like abound and not an input?

JakobBD replied:
not this one (05_inicap/on/preloop.gms)

0UmfHxcvx5J7JoaOhFSs5mncnisTJJ6q replied:
> not this one (05_inicap/on/preloop.gms).

My bad.

> so I could introduce a parameter (as mentioned above), use it in inicap

Probably best.  Some years down the line some poor PhD will wonder why there is a variable that is not variable at all.

LaviniaBaumstark replied:
I agree, introcuding an additonal parameter is teh cleanest way

LaviniaBaumstark commented on scripts/start/calculate_CES_configuration.R:
```diff
@@ -5,7 +5,7 @@
 # |  Contact: remind@pik-potsdam.de
 calculate_CES_configuration <- function(cfg, check = FALSE) {
-    CESstring <- paste0("indu_", cfg$gms$industry,"-",
+    CESstring <- paste0("indu_", cfg$gms$industry,ifelse(cfg$gms$cm_subsec_model_steel=="processes","pbs",""),"-",
```
we will try to get rid of this if-clause as soon as possible, means when thsi will be our default :-)

JakobBD replied:
Independent of which is default, we have to distinguish between the two as long as 1) this is part of the subsectors realization and 2) we retain both options (processes and ces)?

LaviniaBaumstark replied:
sure, I was in teh far future when there are all parts of industry transfered and our default :-)

0UmfHxcvx5J7JoaOhFSs5mncnisTJJ6q reviewed (commented):
> - [ ] All automated model tests pass (FAIL 0 in the output of make test) (**note**: did not run make test, but ran all AMT runs, see below)

Kind of sensible, but maybe worthwhile to include pbs in the test, so it does not get wrecked by somebody else.

Also, run `make check` and incorporate those changes (even the pointless, awful one).

JakobBD commented:
Thanks for your comments! I've addressed them all; in particular, I introduced an intermediate parameter for initial production and moved the fixings to `bounds.gms`. I did an NPi and a PkBudget run to test the changes. Will merge now.
